README.md
# Fiat Methods

Methods for working with fiat prices.

## `getPrices`

Get exchange rates from a rate provider

    {
      from: <From Rate Provider String>
      request: <Request Function> {url}, (err, {statusCode}, body) => {}
      symbols: [<Fiat Symbol String>] // empty defaults to USD
    }

    @returns via cbk or Promise
    {
      tickers: [{
        date: <Rate Updated At ISO 8601 Date String>
        rate: <Exchange Rate in Cents Number>
        ticker: <Ticker Symbol String>
      }]
    }


CHANGELOG.md
# Versions

## 2.0.0

- Require node.js 16 or higher

## 1.0.4

- Add `getPrices` method to get fiat prices


test/rates/test_get_prices.js
const {deepStrictEqual} = require('node:assert').strict;
const {rejects} = require('node:assert').strict;
const test = require('node:test');

const method = require('./../../rates/get_prices');

const makeRequest = (err, r, body) => ({}, cbk) => cbk(err, r, body);
const updatedISO = '2020-01-13T20:13:00+00:00';

const makeArgs = override => {
  const args = {
    from: 'coindesk',
    request: makeRequest(
      null,
      {statusCode: 200},
      {bpi: {USD: {rate_float: 1}}, time: {updatedISO: updatedISO}},
    ),
    symbols: [],
  };

  Object.keys(override).forEach(key => args[key] = override[key]);

  return args;
};

const tests = [
  {
    args: makeArgs({from: undefined}),
    description: 'From is required',
    error: [400, 'ExpectedFromRateProviderToGetPriceRates'],
  },
  {
    args: makeArgs({request: undefined}),
    description: 'Request is required',
    error: [400, 'ExpectedRequestFunctionToGetExchangeRates'],
  },
  {
    args: makeArgs({symbols: undefined}),
    description: 'Symbols are required',
    error: [400, 'ExpectedSymbolsToGetFiatExchangeRates'],
  },
  {
    args: makeArgs({symbols: ['UNKNOWN']}),
    description: 'Known symbols are required',
    error: [400, 'UnsupportedFiatTypeForCoindeskFiatRateLookup'],
  },
  {
    args: makeArgs({}),
    description: 'Rate is returned',
    expected: {rate: 100},
  },
  {
    args: makeArgs({
      from: 'coingecko',
      request: ({}, cbk) => cbk(
        null,
        {statusCode: 200},
        {rates: {eur: {value: 2}, usd: {value: 1}}},
      ),
    }),
    description: 'Batch rates are returned',
    expected: {rate: 100},
  },
];

tests.forEach(({args, description, error, expected}) => {
  return test(description, async () => {
    if (!!error) {
      await rejects(method(args), error, 'Got expected error');
    } else {
      const [{rate}] = (await method(args)).tickers;

      deepStrictEqual(rate, expected.rate, 'Got expected exchange rate');
    }

    return;
  });
});


test/rates/test_get_current_price.js
const {deepStrictEqual} = require('node:assert').strict;
const {rejects} = require('node:assert').strict;
const test = require('node:test');

const method = require('./../../rates/get_current_price');

const makeRequest = (err, r, body) => ({}, cbk) => cbk(err, r, body);
const updatedISO = '2020-01-13T20:13:00+00:00';

const makeArgs = override => {
  const args = {
    currency: 'BTC',
    fiat: 'USD',
    from: 'coindesk',
    request: makeRequest(
      null,
      {statusCode: 200},
      {bpi: {USD: {rate_float: 1}}, time: {updatedISO: updatedISO}},
    )
  };

  Object.keys(override).forEach(key => args[key] = override[key]);

  return args;
};

const tests = [
  {
    args: makeArgs({currency: undefined}),
    description: 'Currency is required',
    error: [400, 'ExpectedCurrencyToGetCurrentPriceFor'],
  },
  {
    args: makeArgs({fiat: undefined}),
    description: 'Fiat is required',
    error: [400, 'ExpectedFiatSymbolToGetCurrentPriceIn'],
  },
  {
    args: makeArgs({from: undefined}),
    description: 'From is required',
    error: [400, 'ExpectedRateProviderToGetCurrentPrice'],
  },
  {
    args: makeArgs({from: 'unknown'}),
    description: 'A known provider is required',
    error: [400, 'UnknownRateProviderToGetCurrentFiatPrice'],
  },
  {
    args: makeArgs({request: undefined}),
    description: 'Request is required',
    error: [400, 'ExpectedRequestFunctionToGetCurrentPrice'],
  },
  {
    args: makeArgs({}),
    description: 'Rate is returned',
    expected: {cents: 100},
  },
];

tests.forEach(({args, description, error, expected}) => {
  return test(description, async () => {
    if (!!error) {
      await rejects(method(args), error, 'Got expected error');

      return;
    }

    const {cents} = await method(args);

    deepStrictEqual(cents, expected.cents, 'Got expected exchange rate');

    return;
  });
});


test/coingecko/test_get_coingecko_rates.js
const {deepStrictEqual} = require('node:assert').strict;
const {rejects} = require('node:assert').strict;
const test = require('node:test');

const {getCoingeckoRates} = require('./../../coingecko');

const updatedISO = '2020-01-13T20:13:00+00:00';

const body = {rates: {eur: {value: 1}, usd: {value: 2}}};

const makeRequest = (err, res) => ({}, cbk) => cbk(err, null, res);

const makeArgs = override => {
  const args = {request: makeRequest(null, body), symbols: ['EUR']};

  Object.keys(override).forEach(key => args[key] = override[key]);

  return args;
};

const tests = [
  {
    args: makeArgs({symbols: undefined}),
    description: 'Symbols are required',
    error: [400, 'ExpectedSymbolsToGetCoingeckoExchangeRates'],
  },
  {
    args: makeArgs({request: undefined}),
    description: 'A request function is required',
    error: [400, 'ExpectedRequestFunctionToGetExchangeRates'],
  },
  {
    args: makeArgs({request: makeRequest('err')}),
    description: 'Request cannot return error',
    error: [503, 'UnexpectedErrorGettingCoingeckoRates', {err: 'err'}],
  },
  {
    args: makeArgs({request: makeRequest(null, null)}),
    description: 'Response must return a body',
    error: [503, 'ExpectedRatesInCoingeckoResponse'],
  },
  {
    args: makeArgs({request: makeRequest(null, {})}),
    description: 'Response must return rates',
    error: [503, 'ExpectedRatesInCoingeckoResponse'],
  },
  {
    args: makeArgs({request: makeRequest(null, {rates: {}})}),
    description: 'Response is expected to have a result',
    error: [404, 'CoingeckoRateLookupSymbolNotFound'],
  },
  {
    args: makeArgs({symbols: []}),
    description: 'Got default exchange rate',
    expected: {
      tickers: [{
        date: '2020-01-13T20:13:00.000Z',
        rate: 200,
        ticker: 'USD',
      }],
    },
  },
  {
    args: makeArgs({}),
    description: 'Got exchange rates',
    expected: {
      tickers: [{
        date: '2020-01-13T20:13:00.000Z',
        rate: 100,
        ticker: 'EUR',
      }],
    },
  },
];

tests.forEach(({args, description, error, expected}) => {
  return test(description, async () => {
    if (!!error) {
      await rejects(getCoingeckoRates(args), error, 'Got expected error');

      return;
    }

    const [expectedTicker] = expected.tickers;
    const [ticker] = (await getCoingeckoRates(args)).tickers;

    deepStrictEqual(!!ticker.date, true, 'Got ticker date');
    deepStrictEqual(ticker.rate, expectedTicker.rate, 'Got expected rate');
    deepStrictEqual(ticker.ticker, expectedTicker.ticker, 'Got symbol');

    return;
  });
});


rates/get_current_price.js
const asyncAuto = require('async/auto');
const {returnResult} = require('asyncjs-util');

const {getCoinbaseCurrentPrice} = require('./../coinbase');
const {getCoindeskCurrentPrice} = require('./../coindesk');

/** Get the current fiat price from a specified rate provider

  {
    currency: <Currency String>
    fiat: <Fiat String>
    from: <From Rate Provider String>
    request: <Request Function>
  }

  @returns via cbk or Promise
  {
    cents: <Cents Per Token Number>
    date: <Updated At ISO 8601 Date String>
  }
*/
module.exports = ({currency, fiat, from, request}, cbk) => {
  return new Promise((resolve, reject) => {
    return asyncAuto({
      // Check arguments
      validate: cbk => {
        if (!currency) {
          return cbk([400, 'ExpectedCurrencyToGetCurrentPriceFor']);
        }

        if (!fiat) {
          return cbk([400, 'ExpectedFiatSymbolToGetCurrentPriceIn']);
        }

        if (!from) {
          return cbk([400, 'ExpectedRateProviderToGetCurrentPrice']);
        }

        if (!request) {
          return cbk([400, 'ExpectedRequestFunctionToGetCurrentPrice']);
        }

        return cbk();
      },

      // Get the current price
      getPrice: ['validate', ({}, cbk) => {
        const providers = {
          coinbase: getCoinbaseCurrentPrice,
          coindesk: getCoindeskCurrentPrice,
        };

        if (!providers[from]) {
          return cbk([400, 'UnknownRateProviderToGetCurrentFiatPrice']);
        }

        return providers[from]({currency, fiat, request}, cbk);
      }],
    },
    returnResult({reject, resolve, of: 'getPrice'}, cbk));
  });
};


coinbase/get_coinbase_current_price.js
const asyncAuto = require('async/auto');
const {returnResult} = require('asyncjs-util');

const api = 'https://api.coinbase.com/';
const centsPerUnit = 100;
const {round} = Math;
const supportedFiats = ['EUR', 'USD'];

/** Get the number of cents for a Bitcoin from Coinbase

  {
    currency: <Currency String>
    fiat: <Fiat String>
    request: <Request Function>
  }

  @returns via cbk or Promise
  {
    cents: <Cents Per Token Number>
    date: <Updated At ISO 8601 Date String>
  }
*/
module.exports = ({currency, fiat, request}, cbk) => {
  return new Promise((resolve, reject) => {
    return asyncAuto({
      // Check arguments
      validate: cbk => {
        if (currency !== 'BTC') {
          return cbk([400, 'ExpectedCurrencyCodeToGetCoinbaseCurrentPrice']);
        }

        if (!supportedFiats.includes(fiat)) {
          return cbk([400, 'ExpectedKnownCurrencyToGetCoinbaseCurrentPrice']);
        }

        if (!request) {
          return cbk([400, 'ExpectedRequestMethodToGetCoinbaseCurrentPrice']);
        }

        return cbk();
      },

      // Get the price data
      getRate: ['validate', ({}, cbk) => {
        return request({
          json: true,
          url: `${api}v2/prices/${currency}-${fiat}/spot`,
        },
        (err, r, body) => {
          if (!!err) {
            return cbk([503, 'UnexpectedErrorGettingCoinbasePrice', {err}]);
          }

          if (!body || !body.data || !body.data.amount) {
            return cbk([503, 'ExpectedCurrencyRateDataFromCoinbase']);
          }

          const cents = round(body.data.amount * centsPerUnit);

          return cbk(null, {cents, date: new Date().toISOString()});
        });
      }],
    },
    returnResult({reject, resolve, of: 'getRate'}, cbk));
  });
};



rates/get_prices.js
const asyncAuto = require('async/auto');
const asyncMap = require('async/map');
const {returnResult} = require('asyncjs-util');

const getCurrentPrice = require('./get_current_price');
const {getCoingeckoRates} = require('./../coingecko');

const currency = 'BTC';
const defaultFiat = 'USD';
const {isArray} = Array;
const uniq = arr => Array.from(new Set(arr));

/** Get exchange rates from a rate provider

  {
    from: <From Rate Provider String>
    request: <Request Function> {url}, (err, {statusCode}, body) => {}
    symbols: [<Fiat Symbol String>] // empty defaults to USD
  }

  @returns via cbk or Promise
  {
    tickers: [{
      date: <Rate Updated At ISO 8601 Date String>
      rate: <Exchange Rate in Cents Number>
      ticker: <Ticker Symbol String>
    }]
  }
*/
module.exports = ({from, request, symbols}, cbk) => {
  return new Promise((resolve, reject) => {
    return asyncAuto({
      // Check arguments
      validate: cbk => {
        if (!from) {
          return cbk([400, 'ExpectedFromRateProviderToGetPriceRates']);
        }

        if (!request) {
          return cbk([400, 'ExpectedRequestFunctionToGetExchangeRates']);
        }

        if (!isArray(symbols)) {
          return cbk([400, 'ExpectedSymbolsToGetFiatExchangeRates']);
        }

        return cbk();
      },

      // Symbols to get
      priceSymbols: ['validate', ({}, cbk) => {
        const prices = [];

        symbols.forEach(n => prices.push(n));

        if (!symbols.length) {
          prices.push(defaultFiat);
        }

        return cbk(null, uniq(prices));
      }],

      // Batch fetch prices
      getBatchPrices: ['priceSymbols', ({priceSymbols}, cbk) => {
        // Exit early when batch rate lookups are supported
        if (from !== 'coingecko') {
          return cbk();
        }

        return getCoingeckoRates({request, symbols: priceSymbols}, cbk);
      }],

      // Fetch all the prices
      getPrices: ['priceSymbols', ({priceSymbols}, cbk) => {
        // Exit early when batch rate lookups are supported
        if (from === 'coingecko') {
          return cbk();
        }

        return asyncMap(priceSymbols, (fiat, cbk) => {
          return getCurrentPrice({
            currency,
            fiat,
            from,
            request,
          },
          (err, res) => {
            if (!!err) {
              return cbk(err);
            }

            return cbk(null, {date: res.date, rate: res.cents, ticker: fiat});
          });
        },
        cbk);
      }],

      // Final set of prices
      prices: [
        'getBatchPrices',
        'getPrices',
        ({getBatchPrices, getPrices}, cbk) =>
      {
        // Exit early when price fetching was batched
        if (!!getBatchPrices) {
          return cbk(null, getBatchPrices);
        }

        return cbk(null, {tickers: getPrices});
      }],
    },
    returnResult({reject, resolve, of: 'prices'}, cbk));
  });
};
